cmake: remove usage of zephyr_library_ifdef

AdithyaBaglody opened the pull request:
All instances of the cmake macro zephyr_library_ifdef will be removed in this PR.

[SEBO] EDIT: Fixes #7565

AdithyaBaglody committed:
arch: arm: add_subdirectory shouldn't depend on a hidden kconfig.

Previously the directory core/mpu/ was getting included on a hidden
kconfig. Now this has been replaced with the Kconfig
CONFIG_ARC_CORE_MPU.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
arch: arm: add_subdirectory shouldn't depend on a hidden kconfig.

Previously the directory core/cortex/mpu/ was getting included
on a hidden kconfig. Now this has been replaced with the Kconfig
CONFIG_ARM_CORE_MPU.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
arch: arm/arc: Remove usage of zephyr_library_ifdef.

Phasing out the usage of this cmake macro from the arch folder.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
boards: arm: cc2650: Remove CONFIG_FLASH from defconfig.

Fixed the incorrect usage of CONFIG_FLASH in the defconfig.
This CONFIG_FLASH is used to enable the flash driver only.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
boards: arm: cc3220sf: Remove CONFIG_FLASH from defconfig.

Fixed the incorrect usage of CONFIG_FLASH in the defconfig.
This CONFIG_FLASH is used to enable the flash driver only.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
boards: arm: msp_exp432p401r: Remove CONFIG_FLASH from defconfig.

Fixed the incorrect usage of CONFIG_FLASH in the defconfig.
This CONFIG_FLASH is used to enable the flash driver only.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

SebastianBoe commented:
Why DNM?

SebastianBoe commented:
The definition of "zephyr_library_ifdef" should also be removed in this PR.

AdithyaBaglody commented:
I am adding a few more commits to this PR. 
I want to check how the CI behaves before I push all of them. 
There is also an issue #7663 which I  want to check before pushing anything.

AdithyaBaglody committed:
drivers: sensors: Remove usage of zephyr_library_ifdef

Phasing out the usage of this cmake macro from the sensor folder.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
drivers: flash: Remove usage of zephyr_library_ifdef

Phasing out the usage of this cmake macro from the drivers/flash
folder.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

galak commented:
Why are we removing it, that doesn't seem to be explained anywhere.

nashif commented:
> Why are we removing it, that doesn't seem to be explained anywhere.

#7565

SebastianBoe commented:
> Why are we removing it, that doesn't seem to be explained anywhere.

Because that would resolve

https://github.com/zephyrproject-rtos/zephyr/issues/7565

AdithyaBaglody commented:
@galak the commit which explain the reason for removing the cmake macro hasn't been pushed yet. will be sent in the subsequent patches.

GAnthony commented on boards/arm/cc3220sf_launchxl/cc3220sf_launchxl_defconfig:
```diff
@@ -4,7 +4,6 @@ CONFIG_SOC_FAMILY_TISIMPLELINK=y
 CONFIG_SOC_SERIES_CC32XX=y
 CONFIG_SOC_CC3220SF=y
 CONFIG_CORTEX_M_SYSTICK=y
-CONFIG_FLASH=y
```
I validated this for cc3220sf.  Looks good.

AdithyaBaglody replied:
Thanks for the confirmation.

AdithyaBaglody committed:
drivers: serial: Remove usage of zephyr_library_ifdef

Phasing out the usage of this cmake macro from the drivers/serial
folder.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
drivers: entropy: Remove usage of zephyr_library_ifdef

Phasing out the usage of this cmake macro from the drivers/entropy
folder.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
cmake: extensions: Remove macro zephyr_library_ifdef.

This cmake macro was changing the control flow of the cmake files.
It is not advisable to modify the callee control flow from inside
a cmake macro.
Hence removing all instances of this macro from the build system.

Fixes: GH-7565

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody commented:
Removing the DNM since this fixes #7565 
For the issue #7663 I will create a new PR when I get the solution.

SebastianBoe commented on drivers/CMakeLists.txt:
```diff
@@ -23,16 +22,17 @@ add_subdirectory_if_kconfig(pinmux)
 add_subdirectory_ifdef(CONFIG_NETWORKING       net)
 add_subdirectory_ifdef(CONFIG_NET_L2_ETHERNET  ethernet)
-add_subdirectory_ifdef(CONFIG_ENTROPY_GENERATOR entropy)
+add_subdirectory_ifdef(CONFIG_ENTROPY_HAS_DRIVER entropy)
```
Will changing these semantics cause entropy drivers, that used to be pruned, to now be included into applications?

Is there any way of proving yes / no. E.g. by running sanitycheck and comparing sizes before and after.

The reason I suspect there is a risk of this is that this build script used to say:

"Include the entropy driver when the entropy generator is enabled."

But now it says:

"Include the entropy driver when the entropy generator is supported."

AdithyaBaglody replied:
@SebastianBoe I understand the issue here. But this Kconfig "entropy has driver" will only be enabled if entropy_generator is enabled. So we will maintain the same subdirectory include condition as before.

SebastianBoe replied:
I see, then this is safe.
